libraries/botbuilder-dialogs-adaptive/src/adaptiveDialogBot.ts

```diff
@@ -0,0 +1,129 @@
+import { languageGeneratorKey, languageGeneratorManagerKey, languagePolicyKey } from './languageGeneratorExtensions';
+import { resourceExplorerKey } from './resourceExtensions';
+import { skillConversationIdFactoryKey } from './skillExtensions';
+
+import {
+    ActivityTypes,
+    Bot,
+    BotCallbackHandlerKey,
+    BotTelemetryClient,
+    BotTelemetryClientKey,
+    ConversationState,
+    SkillConversationIdFactoryBase,
+    TurnContext,
+    UserState,
+} from 'botbuilder';
+
+import {
+    LanguageGeneratorManager,
+    ResourceMultiLanguageGenerator,
+    TemplateEngineLanguageGenerator,
+} from './generators';
+
+export class AdaptiveDialogBot implements Bot {
+    private static readonly languageGeneratorManagers = new Map<ResourceExplorer, LanguageGeneratorManager>();
+
+    private readonly lazyRootDialog: () => Dialog;
+
+    constructor(
+        private readonly adaptiveDialogId: string,
+        private readonly languageGeneratorId: string,
+        private readonly resourceExplorer: ResourceExplorer,
+        private readonly conversationState: ConversationState,
+        private readonly userState: UserState,
+        private readonly skillConversationIdFactoryBase: SkillConversationIdFactoryBase,
+        private readonly languagePolicy: LanguagePolicy,
+        private readonly botFrameworkAuthentication: BotFrameworkAuthentication,
+        private readonly telemetryClient: BotTelemetryClient,
+        private readonly memoryScopes: MemoryScope[] = [],
+        private readonly pathResolvers: PathResolver[] = [],
+        private readonly dialogs: Dialog[] = []
+    ) {
+        this.lazyRootDialog = memoize(() => this.createDialog());
+    }
+
+    async onTurn(context: TurnContext): Promise<void> {
+        const botFrameworkClient = this.botFrameworkAuthentication.createBotFrameworkClient();
+
+        // Set up the TurnState the Dialog is expecting.
+        await this.setUpTurnState(context, botFrameworkClient);
+
+        // Load the Dialog from the ResourceExplorer - the actual load should only happen once.
+        const rootDialog = this.lazyRootDialog();
+
+        // Run the dialog.
+        await runDialog(
+            rootDialog,
+            context,
+            context.turnState.get<ConversationState>('ConversationState').createProperty('dialogState')
+        );
+
+        // Save any updates that have been made.
+        await context.turnState.get<ConversationState>('ConversationState').saveChanges(context);
+        await context.turnState.get<UserState>('UserState').saveChanges(context);
+    }
+
+    private async setUpTurnState(context: TurnContext, botFrameworkClient: BotFrameworkClient): Promise<void> {
+        context.turnState.set(BotFrameworkClientKey, botFrameworkClient);
+        context.turnState.set(skillConversationIdFactoryKey, this.skillConversationIdFactoryBase);
+        context.turnState.set('ConversationState', this.conversationState);
```

Comment:
nit: These appear between 2 and 4 times, can we define constants like we did for other keys? if not public, would be a nice excuse to have all the *key constant variable names have the same casing, i.e. `skillConversationIdFactoryKey` vs `BotFrameworkClientKey`

Reply:
I noticed this as I was writing it and started making the changes. However, there are so many areas this touches (or *should* touch) that it really warrants its own PR. Unfortunately, I'm not sure we can change the casing to match without it being a breaking change...probably some of them are fine, though.

Reply:
I extracted constants for these, I'm open to feedback about how I set it up. The constants are defined alongside the types they refer to since that was the least impactful with respect to imports/dependencies (i.e. anywhere that imports `UserState` already has the appropriate `package.json` dependencies).